### PR TITLE
Fixes #21443 - Create PuppetEnv without an internal proxy

### PR DESCRIPTION
**Pull request by sean797**

### CODE REVIEW

File: app/lib/actions/katello/capsule_content/sync.rb [hunk 1]
```diff
@@ -70,7 +70,6 @@ def sync_repos_to_capsule(capsule_content, repository_ids, skip_metadata_check)
                 if skip_metadata_check
                   plan_action(Katello::Repository::MetadataGenerate,
                               repo,
-                              capsule_id: capsule_content.capsule.id,
```
Comment (sean797) [inline app/lib/actions/katello/capsule_content/sync.rb]: Removed this as `Katello::Repository::MetadataGenerate` doesnt seem to used a `capsule_id` option (not related to this change), maybe that's another bug as I think currently a 'forced' smart proxy sync will regenerate metadata on the "master" pulp server?

File: app/lib/actions/katello/content_view_puppet_environment/create.rb [hunk 2]
```diff
@@ -23,7 +21,7 @@ def plan(puppet_environment, clone = false)
 
             # when creating a clone, the following actions are handled by the
             # publish/promote process
-            unless clone
+            unless clone && !puppet_path
```
Comment (ehelms) [inline app/lib/actions/katello/content_view_puppet_environment/create.rb]: I dislike compound unless, mind changing this to an `if` statement for easier readability? Now that its compound I don't think linting will yell at you.

File: app/lib/actions/katello/content_view_puppet_environment/create.rb [hunk 3]
```diff
@@ -4,15 +4,13 @@ module ContentViewPuppetEnvironment
       class Create < Actions::EntryAction
         def plan(puppet_environment, clone = false)
           internal_capsule = SmartProxy.default_capsule
-          fail _("Content View %s  cannot be published without an internal capsule." % puppet_environment.name) unless internal_capsule
+          puppet_path = internal_capsule.present? ? puppet_environment.generate_puppet_path(internal_capsule) : nil
```
Comment (ehelms) [inline app/lib/actions/katello/content_view_puppet_environment/create.rb]: If this is nil, what happens in the Repository::Create action below given no path would be specified?

**Comment by ehelms** [comment]
I think from my comment you'll get that what I am asking to help clarify from you on this change is, if the puppet path is not being generated by using the default smart proxy as the source of truth, then where does that come from to ensure that Puppet environments wind up in the right place?

File: app/lib/actions/katello/content_view_puppet_environment/create.rb (hunk 3 above)
Comment (sean797) [reply app/lib/actions/katello/content_view_puppet_environment/create.rb]: Without a path we already create it without the `puppet_install_distributor`. Which is what we want when the Pulp nodes isn't also a Puppet Master.
See https://github.com/Katello/katello/blob/master/app/lib/actions/pulp/repository/create.rb#L120

**Comment by sean797** [comment]
When we Sync to a Smart Proxy Capsule we get the path from that Proxy, see https://github.com/Katello/katello/blob/master/app/lib/actions/katello/capsule_content/create_repos.rb#L54. `ContentViewPuppetEnvironment::Create` is only used to create on the 'Master' Pulp server (the one in settings.yml)

File: app/lib/actions/katello/content_view_puppet_environment/create.rb (hunk 3 above)
Comment (ehelms) [reply app/lib/actions/katello/content_view_puppet_environment/create.rb]: For the server side case, could this lead to some confusion for users? Say for some reason, whether it be the server does not have an internal proxy that it knows about or a user did not install a puppet master/server, and they add puppet modules to their CV this won't generate the output from that anywhere. Now perhaps that is to be expected given no Puppet server/master is present, but I could see it confusing people without any output or warning.

Still wrapping my head around this but I am wondering:

 1) What happens if you  move the default proxy off to another box co-located with Puppet master?

Comment (sean797) [reply app/lib/actions/katello/content_view_puppet_environment/create.rb]: I don't think that is confusing, if you have a separate Pulp 'master' without a foreman-proxy(or Puppet) you don't need or want `puppet_install_distributor`.

> What happens if you move the default proxy off to another box co-located with Puppet master?

No change as, https://github.com/Katello/katello/blob/master/app/models/katello/concerns/smart_proxy_extensions.rb#L56 will still return it.

This change is purely to remove the hard requirement between foreman-proxy and Pulp master, obviously without foreman-proxy you don't get disk usage reporting ect.. but it makes the app a lot more flexible.

**Comment by sean797** [comment]
@ehelms are you okay with saying "If you want monitoring you need to installer foreman-proxy with Pulp", as this doesn't change the default way we install I think its okay?

**Comment by sean797** [comment]
Rebased. I assume we are ignoring the "katello" build that's failing since "rails-5" appears to run katello tests?

**Comment by beav** [comment]
@sean797 ya I think it's good now w.r.t. automated tests.

@ehelms can you take a look at https://github.com/Katello/katello/pull/7029#issuecomment-345866626 ?

**Comment by beav** [comment]
@ehelms please check out https://github.com/Katello/katello/pull/7029#issuecomment-345866626, PR is waiting on feedback

**Comment by ehelms** [comment]
Thanks for keeping an eye on this @beav. 

There have been discussions around the intent of a Smart Proxy and what our requirements around it's usage ought to be with respect to backends. There's an RFC discussion that would essentially undo this entire concept by making a Smart Proxy required and all connection information derived from it. I think we need to fully settle that discussion before making this change, this way unfortunately.

**Comment by johnsonm325** [comment]
@ehelms It's been a month and a half since the last update here, so I wasn't sure if the RFC discussion had taken place and any decision made on what to do with this PR. If we're moving forward, @sean797 there is a conflict to be resolved.

**Comment by johnsonm325** [comment]
@ehelms @sean797 Following up here again to see if there is any progress. Let me know if there's something you need from me or the katello team.

**Comment by johnsonm325** [comment]
@ehelms @sean797 Following up here again on this one to see how and if we can move the forward.

**Comment by sean797** [comment]
I think this effectively conflicts with https://github.com/Katello/katello/pull/7470? IIRC I don't think there was a clear resolution from that discussion? Though I can't find it on discourse now :-1:

**Comment by ekohl** [comment]
@sean797 maybe you're referring to [capabilities](https://community.theforeman.org/t/exposing-capabilities-in-the-smart-proxy/9860/8) which @jlsherrill is taking over.

**Comment by sean797** [comment]
Found it! https://community.theforeman.org/t/how-we-think-about-and-use-smart-proxies-architecturally/10231/31 me++

**Comment by sean797** [comment]
Closing, see https://github.com/Katello/katello/pull/7470#issuecomment-545990972

**Pull request closed without merging**